lib/srv/discovery/status.go

```diff
@@ -0,0 +1,120 @@
+ */
+
+package discovery
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/gravitational/trace"
+
+	discoveryconfigv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/discoveryconfig/v1"
+	"github.com/gravitational/teleport/api/types/discoveryconfig"
+	aws_sync "github.com/gravitational/teleport/lib/srv/discovery/fetchers/aws-sync"
+)
+
+// updateDiscoveryConfigStatus updates the DiscoveryConfig Status field with the current in-memory status.
+// The status will be updated with the following matchers:
+// - AWS Sync (TAG) status
+func (s *Server) updateDiscoveryConfigStatus(discoveryConfigName string) {
+	discoveryConfigStatus := discoveryconfig.Status{
+		State:        discoveryconfigv1.DiscoveryConfigState_DISCOVERY_CONFIG_STATE_SYNCING.String(),
+		LastSyncTime: s.clock.Now(),
+	}
+
+	// Merge AWS Sync (TAG) status
+	discoveryConfigStatus = s.awsSyncStatus.mergeIntoGlobalStatus(discoveryConfigName, discoveryConfigStatus)
+
+	ctx, cancel := context.WithTimeout(s.ctx, 5*time.Second)
+	defer cancel()
+
+	_, err := s.AccessPoint.UpdateDiscoveryConfigStatus(ctx, discoveryConfigName, discoveryConfigStatus)
+	switch {
+	case trace.IsNotImplemented(err):
+		s.Log.Warn("UpdateDiscoveryConfigStatus method is not implemented in Auth Server. Please upgrade it to a recent version.")
+	case err != nil:
+		s.Log.WithError(err).Infof("Error updating discovery config %q status", discoveryConfigName)
+	}
+}
+
+func newAWSSyncStatus() awsSyncStatus {
+	return awsSyncStatus{
+		awsSyncStatus: make(map[string]awsSyncResult),
+	}
+}
+
+type awsSyncStatus struct {
+	mu            sync.RWMutex
+	awsSyncStatus map[string]awsSyncResult
+}
+
+type awsSyncResult struct {
+	state               string
+	errorMessage        *string
+	lastSyncTime        time.Time
+	discoveredResources uint64
+}
+
+func (d *awsSyncStatus) upsertStatus(discoveryConfig string, result awsSyncResult) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	if d.awsSyncStatus == nil {
+		d.awsSyncStatus = make(map[string]awsSyncResult)
+	}
+	d.awsSyncStatus[discoveryConfig] = result
+}
+
+func (d *awsSyncStatus) mergeIntoGlobalStatus(discoveryConfig string, existingStatus discoveryconfig.Status) discoveryconfig.Status {
+	d.mu.RLock()
+	defer d.mu.RUnlock()
+
+	awsStatus, found := d.awsSyncStatus[discoveryConfig]
+	if !found {
+		return existingStatus
+	}
+
+	existingStatus.DiscoveredResources = existingStatus.DiscoveredResources + awsStatus.discoveredResources
+	existingStatus.ErrorMessage = awsStatus.errorMessage
+	existingStatus.State = awsStatus.state
+	existingStatus.LastSyncTime = awsStatus.lastSyncTime
+
+	return existingStatus
+}
+
+// updateAWSSyncDiscoveryConfigStatus updates the status for each DiscoveryConfig that originated the Fetchers.
+// It updates the internal state and updates the Status in the cluster.
+func (s *Server) updateAWSSyncDiscoveryConfigStatus(fetchers []aws_sync.AWSSync, pushErr error, preRun bool) error {
```

Comment:
nit: Maybe keep this and Server.updateDiscoveryConfigStatus together?

Reply:
This one is specifically for AWS Sync Fetchers, which are not always present in a DiscoveryConfig.

I think having them as two separate methods helps on extensibility, specifically when adding status for other Fetchers.

Reply:
I don't mean to merge the methods, just to move them next to each other in the file. These are all *Server methods, keeping them close together helps people reading the code realize they exist and relate to each other.

Reply:
They are in `status.go` file.
I'm including everything related to updating the Status of a DiscoveryConfig in this file.
Can you please check again?

Reply:
I think the comment is just outdated, resolving.